### PR TITLE
Reuse request body with net/http.Request.GetBody if available

### PR DESCRIPTION
**Pull request by magnusbaeck**
Commit 3b7925a addressed the problem where Transport.RoundTrip didn't work with a non-nil body since the original io.Reader had been emptied by the first request (i.e. the second request got an immediate EOF) but did so by reading the whole payload into RAM. That's totally fine for small requests but didn't scale with large payloads and/or high concurrency.

We address this by paying attention to net/http.Request.GetBody (introduced in Go 1.8), which if non-nil contains the address of a function that produces a fresh io.Reader. If that field is unset we fall back to the old behavior of reading the payload to RAM.

I tested this with existing code for posting a 50 MB payload, and the process's runtime.MemStats.HeapSys decreased from 200 MB to 63 MB.

### CODE REVIEW

File: digest.go
Comment (edaniels) [inline digest.go:281]: can you refactor this to up above and check first if `req.Body != nil` and then have an if/else for when GetBody is nil or not?

**Review by edaniels — CHANGES_REQUESTED** [review]
Just one requested change for organization but otherwise LGTM after that!

File: digest.go
Comment (magnusbaeck) [reply digest.go:281]: The reason I put the GetBody call way down here was to avoid req2.Body.Close calls before each error return (lines 260, 263, 271, and 278). Are you sure about paying that price to be able to keep all body-related code in one place?

Comment (edaniels) [reply digest.go:281]: that's fair, carry on :)

**Review by edaniels — APPROVED** [review]

**Comment by edaniels** [comment]
@magnusbaeck, I just need you to GPG sign your commit before I can merge it.

**Comment by magnusbaeck** [comment]
@edaniels GPG signature in place so we should be ready to merge. Thanks for the quick turnaround time!

**Pull request merged**
